torchvision/prototype/transforms/_transform.py

```diff
@@ -37,9 +38,32 @@ def forward(self, *inputs: Any) -> Any:
 
         params = self._get_params(flat_inputs)
 
-        flat_outputs = [
-            self._transform(inpt, params) if check_type(inpt, self._transformed_types) else inpt for inpt in flat_inputs
-        ]
+        # Below is a heuristic on how to deal with simple tensor inputs:
+        # 1. Simple tensors, i.e. tensors that are not a datapoint, are passed through if there is an explicit image
+        #    (`datapoints.Image` or `PIL.Image.Image`) or video (`datapoints.Video`) in the sample.
+        # 2. If there is no explicit image or video in the sample, only the first encountered simple tensor is
+        #    transformed as image, while the rest is passed through. The order is defined by the returned `flat_inputs`
+        #    of `tree_flatten`, which recurses depth-first through the input.
+        #
+        # This heuristic stems from two requirements:
+        # 1. We need to keep BC for single input simple tensors and treat them as images.
+        # 2. We don't want to treat all simple tensors as images, because some datasets like `CelebA` or `Widerface`
+        #    return supplemental numerical data as tensors that cannot be transformed as images.
+        #
+        # The heuristic should work well for most people in practice. The only case where it doesn't is if someone
+        # tries to transform multiple simple tensors at the same time, expecting them all to be treated as images.
+        # However, this case wasn't supported by transforms v1 either, so there is no BC concern.
+        flat_outputs = []
+        simple_tensor_transformed = has_any(flat_inputs, datapoints.Image, datapoints.Video, PIL.Image.Image)
+        for inpt in flat_inputs:
+            if is_simple_tensor(inpt):
+                if simple_tensor_transformed:
+                    flat_outputs.append(inpt)
+                    continue
+
+                simple_tensor_transformed = True
```

Comment:
The `continue` seems unnecessary. Remove it or use as below?

```suggestion
                if simple_tensor_transformed:
                    flat_outputs.append(inpt)
                else:
                    simple_tensor_transformed = True
```

Reply:
It is not. Note that we are in a double `if`. The outer checks if we are dealing with a simple tensor. The inner then verifies if we want to transform it. If we don't want to, we need start the loop from the beginning, or otherwise we are going to hit `flat_outputs.append(...)` twice. 

I'm not saying that we can't simplify here, but your suggestion does not work. I'm gonna push extensive tests soon. Happy for you to try to improve on my version.

Reply:
I needed to refactor some more. PTAL if this is still relevant.